cmd/kubeadm/app/phases/certs/certs.go

```diff
@@ -545,6 +678,63 @@ func getAltNames(cfg *kubeadmapi.MasterConfiguration) (*certutil.AltNames, error
 			altNames.IPs = append(altNames.IPs, ip)
 		} else if len(validation.IsDNS1123Subdomain(altname)) == 0 {
 			altNames.DNSNames = append(altNames.DNSNames, altname)
+		} else {
+			fmt.Printf("[certificates] WARNING: '%s' was not added to the '%s' SAN, because it is not a valid IP or RFC-1123 compliant DNS entry\n", altname, kubeadmconstants.APIServerCertName)
+		}
+	}
+
+	return altNames, nil
+}
+
+// getEtcdAltNames builds an AltNames object for generating the etcd server certificate
+// localhost is included in the SAN since this is the interface the etcd static pod listens on
+// hostname and API.AdvertiseAddress are excluded since etcd does not listen on this interface by default
+// the user can override the listen address with Etcd.ExtraArgs and add SANs with EtcdCertSANs
+func getEtcdAltNames(cfg *kubeadmapi.MasterConfiguration) (*certutil.AltNames, error) {
+	altNames := &certutil.AltNames{
+		DNSNames: []string{"localhost"},
+		IPs:      []net.IP{net.IPv4(127, 0, 0, 1)},
+	}
+
+	// adds additional SAN
+	for _, altname := range cfg.EtcdCertSANs {
+		if ip := net.ParseIP(altname); ip != nil {
+			altNames.IPs = append(altNames.IPs, ip)
+		} else if len(validation.IsDNS1123Subdomain(altname)) == 0 {
+			altNames.DNSNames = append(altNames.DNSNames, altname)
+		} else {
+			fmt.Printf("[certificates] WARNING: '%s' was not added to the '%s' SAN, because it is not a valid IP or RFC-1123 compliant DNS entry\n", altname, kubeadmconstants.EtcdCertName)
```

Comment:
let's define this error once

Reply:
thanks for the suggestion :+1: this is done